[mlir][python] expose LLVMStructType API

ftynse opened the pull request:
Expose the API for constructing and inspecting StructTypes from the LLVM dialect. Separate constructor methods are used instead of overloads for better readability, similarly to IntegerType.

nicolasvasilache commented on mlir/include/mlir-c/Dialect/LLVM.h:
```diff
@@ -34,11 +34,71 @@ MLIR_CAPI_EXPORTED MlirType
+MLIR_CAPI_EXPORTED MlirStringRef mlirLLVMStructTypeGetIdentifier(MlirType type);
+
+/// Returns `true` is the struct is explicitly opaque (will not have a body) or
+/// unitiniazlied (will eventually have a body).
```
nit: typo

nicolasvasilache commented on mlir/lib/CAPI/Dialect/LLVM.cpp:
```diff
@@ -36,11 +36,75 @@ MlirType mlirLLVMFunctionTypeGet(MlirType resultType, intptr_t nArgumentTypes,
+  return wrap(
+      cast<LLVM::LLVMStructType>(unwrap(structType))
+          .setBody(unwrapList(nFieldTypes, fieldTypes, fields), isPacked));
+}
```
nit: nl

nicolasvasilache commented on mlir/lib/Bindings/Python/DialectLLVM.cpp:
```diff
@@ -0,0 +1,145 @@
+  m.doc() = "MLIR LLVM Dialect";
+
+  populateDialectLLVMSubmodule(m);
+}
```
nit:nl

nicolasvasilache commented on mlir/lib/Bindings/Python/DialectLLVM.cpp:
```diff
@@ -0,0 +1,145 @@
+using namespace mlir::python::adaptors;
+
+namespace {
+/// Standalone RAII scope guard. We don't want to depend on the LLVM support
```
+1 thanks for being mindful of this

nicolasvasilache commented on mlir/test/python/dialects/llvm.py:
```diff
@@ -15,6 +15,90 @@ def constructAndPrintInModule(f):
+    except ValueError as e:
+        pass
+    else:
+        assert False, "expected exception not raisr"
```
nit: typo

nicolasvasilache reviewed (approved):
Nice, thanks!

ftynse committed:
[mlir] expose LLVMStructType API

Expose the API for constructing and inspecting StructTypes from the LLVM
dialect. Separate constructor methods are used instead of overloads for
better readability, similarly to IntegerType.
